### PR TITLE
ARROW-2790: [C++] Buffers can contain uninitialized memory

### PR DESCRIPTION
**Pull request by alendit**

### CODE REVIEW

**Comment by pitrou** [comment]
HIgh-level question: do we care about the potential performance impact? I'd say no, but I'd like to have other people's feeling on this. @xhochy

**Comment by xhochy** [comment]
Actually we do and I always dislike the zeroing of memory. This is a feature though that we could improve by better use of the underlying allocator.  When we push the zeroing logic to the allocator interface, we could call `memset` always in the case when we use the POSIX API. In the case of `jemalloc`, we should set the `MALLOCX_ZERO` flag on the allocator calls and then `jemalloc` can determine if it need to zero the memory or whether the OS already provided a zero-ed page (the OS does this once when a page is newly assigned to a process).

**Comment by xhochy** [comment]
TL;DR: +1 for this PR, we should do a general follow-up some time later.

**Comment by pitrou** [comment]
(reference issue for better collaboration with the allocator is https://issues.apache.org/jira/browse/ARROW-665)

**Comment by alendit** [comment]
The lack of alligned calloc is a real shame. malloc+memset are most likely slower than this solution, but would save A LOT of complexity. 

What is the policy regarding jemalloc? Is it an experimental feature, "first class" supported or maybe even recommended?

**Comment by xhochy** [comment]
`jemalloc` is first class and recommended. We have it activated in our `pip` and `conda` packages by default.

**Comment by alendit** [comment]
Fixing the issue in the allocator is nice, because it's 5 LOC with almost no room for something to go wrong. 

I was surprised by the impact of `MALLOCX_ZERO`, though. This might be because `builder-benchmark` is a pathological case for it, since it repeatedly allocated and deallocates the memory, rarely requesting new zero-pages from the system, so take it with a grain of salt: https://pastebin.com/Kygf5BVp. The impact of memsetting is pretty similar. 

I'm not sure what's the best option in this case.

**Comment by pitrou** [comment]
Zero pages are free only if they stay zeroes (i.e. they remain mapped to `/dev/zero` or something). As soon as you mutate even one byte in a zero page, I suspect a CoW-like mechanism has to remap the page to a new physical memory area, and do an actual `memset()` (or the in-kernel equivalent) to preserve the zeroing.

I think this PR has the right approach: only zero bytes that you know won't be initialized afterwards. It's a bit more complex, but far more efficient.

**Comment by wesm** [comment]
It's the July 4 holiday in the US so I'm not working very much today, but I'd like to review this before it gets merged

**Comment by pitrou** [comment]
So would I, which is why I self-requested a review :-)

**Comment by pcmoritz** [comment]
Recently we have been hitting this problem in Ray, i.e. generated arrow objects are not deterministic any more.

How about we introduce a new macro that writes the padding zeros (which can be switched off)? Then we can let the user decide if they want to use os-level zeroing in JEMALLOC or the mechanism introduced in this PR and can in the former case switch off the explicit zeroing without additional overhead.

For Ray I think I'm going to try to go with the JEMALLOC zeroing approach.

**Comment by wesm** [comment]
We can take a "nuke it from orbit" approach but it is would be good to isolate the source of the non-determinism and have some tests for it. For the tests we could make an allocator that always returns buffers containing nonzero bytes

**Comment by alendit** [comment]
Hi Wes,

the tests in 138b357 and dc7c316 are pretty much standalone, I've used them with both "zero all" and "finegrained zeroing". The "zero all" is in [this branch](https://github.com/alendit/arrow/tree/fix-uninitialized-alloc). It's way simpler, but, as I said before, has non-negligible performance costs, even on jemalloc. On plain posix, where you have to eagerly zero the whole range, it can be even worse, especially for bigger buffer, when they exponentially grow in size.

But the simplicity might still be worth it. Or maybe have it behind a compile flag. Or maybe even add a custom allocator attribute which can checked at runtime to decide if the memory is already zeroed and have memory pool be runtime pluggable  (might be A BIT overengineered ;))

**Comment by pcmoritz** [comment]
I looked into the non-determinism problems with Ray now, and conclude it was introduced in https://github.com/apache/arrow/commit/d19089e5a1179c654cb30df62433698c93e3cd60 and it is not fixed by this PR, but is fixed by https://github.com/alendit/arrow/commit/98f681012bd4039f0bc6995d2229553e37e9f11d.

**Comment by wesm** [comment]
@pcmoritz that makes sense, I reviewed that patch. I didn't think about the implications of the uninitialized padding bytes, and now we have it. We should definitely zero the padding; I would say unconditionally. We can shrink the size of a buffer without reallocating. So if we have a ton of unused capacity, we can shrink to the desired padding (multiple of 64 bytes) and pad any uninitialized bytes.

**Comment by wesm** [comment]
Let's try to fix without zeroing memory in the allocator. It would be a shame to write memory twice when we don't have to

**Comment by pcmoritz** [comment]
Yeah, that sounds good. So I guess we need to find what was changed by https://github.com/apache/arrow/commit/d19089e5a1179c654cb30df62433698c93e3cd60 but not zeroed by this PR. If you have a hunch let me know!

**Comment by alendit** [comment]
Hi @pcmoritz,

could you test with new commits (753f1473)? There was a bit uninitialized padding from `BitUtil::BytesToBits`, maybe with was the final culprit. 

If it still doesn't work, could you point me to the ray code where the non-deterministic arrays are generated?

File: cpp/src/arrow/array-test.cc
Comment (pitrou) [inline cpp/src/arrow/array-test.cc]: Please put this block after the `using` statements below.

File: cpp/src/arrow/array-test.cc
Comment (pitrou) [inline cpp/src/arrow/array-test.cc]: Could you add a comment / docstring here?

File: cpp/src/arrow/array-test.cc
Comment (pitrou) [inline cpp/src/arrow/array-test.cc]: Why is the cast necessary?

File: cpp/src/arrow/array-test.cc
Comment (pitrou) [inline cpp/src/arrow/array-test.cc]: Sometimes we only call `TestInitialized`, sometimes we call `IsZeroPadded`, sometimes both... is there a rationale? Otherwise I'd rather call them both always.

File: cpp/src/arrow/builder.h
Comment (pitrou) [inline cpp/src/arrow/builder.h]: Wait... If `valid_bytes` contains non-null entries, and I already filled in those non-null entries, this will overwrite them with zeros?
I think at the very least this should be mentioned in the docstring.

File: cpp/src/arrow/array-test.cc
Comment (pitrou) [inline cpp/src/arrow/array-test.cc]: ASSERT_EQ

File: cpp/src/arrow/array-test.cc
Comment (pitrou) [inline cpp/src/arrow/array-test.cc]: ASSERT_EQ

File: cpp/src/arrow/builder.h
Comment (pitrou) [inline cpp/src/arrow/builder.h]: `size_t` needn't a `std::` prefix.

File: cpp/src/arrow/builder.h
Comment (pitrou) [inline cpp/src/arrow/builder.h]: `size_t` needn't a `std::` prefix.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/array-test.cc
Comment (alendit) [reply cpp/src/arrow/array-test.cc]: Left-over from the older implementation, I've removed it.

File: cpp/src/arrow/array-test.cc
Comment (alendit) [reply cpp/src/arrow/array-test.cc]: You're right, I've changed it in the commits I've just pushed.

File: cpp/src/arrow/builder.h
Comment (alendit) [reply cpp/src/arrow/builder.h]: Yeah, I wasn't sure about the intended semantics of `AppendNulls`. Looking at the only use case in the orc adapter, it doesn't cause problems, because it's called before the data is set, but I see how it can be a nasty surprise.

The problem is that we have no way of knowing if the data at the non-null place is uninitialized or intentionally set, so I'll go with a warning in a comment.

File: cpp/src/arrow/array-test.cc
Comment (alendit) [reply cpp/src/arrow/array-test.cc]: Oh, at this exact place it's actually intended: for primitive / adaptive int builders we have separate tests for uninitialized memory and for zero-padding. For more complex builders we check it both at places like `ValidateBasicListArray`.

**Comment by alendit** [comment]
Continuing the discussion about AppendNulls: is it even needed, at least in the form with custom `valid_bytes` instead of just length? We already have append values with basically the same functionality, and with clear semantics. I'm not sure how I feel about [orc adapter](https://github.com/apache/arrow/blob/master/cpp/src/arrow/adapters/orc/adapter.cc#L530): couldn't you do the same thing with `AppendValues` (this code predates this API call, most likely, but still)? With a mutable access to builder's data buffers one can wreck all sorts of stuff. I'd go as far as to return a immutable slice from the `data()` of the builder, but maybe I overlook some legitimate use case.

**Comment by pcmoritz** [comment]
Thanks @alendit! I'm testing your PR here: https://github.com/ray-project/ray/pull/2371

If it still fails I'm going to extract the relevant object(s) and make a regression test for arrow out of it so it's easier to investigate where the problem is.

**Comment by wesm** [comment]
If you can somehow create a unit test that would be super helpful

**Comment by pcmoritz** [comment]
Sure thing, I added a test here: https://github.com/apache/arrow/pull/2230

Seems like there is still non-determinism with this PR though (for commit 3697fcf597a7434344bfb3475d2fa0bc6e3bf0a1)

**Comment by wesm** [comment]
@pcmoritz thanks, I'm going to go ahead and cherry-pick that commit into this PR so we can address the issue all at once

**Review by wesm — COMMENTED** [review]
Some comments. I am probably going to push some commits to this branch and fix some of the other buffers that aren't having their padding zero'd (lists, structs, unions) that are impacting the Ray test

**Comment by alendit** [comment]
Hi Wes,

I'm currently expremineting with having a "zero_padded_" attribute at buffer level which is checked every time data() is called (i.e. you get a warning if you get data of an unzeroed buffer). I'm still not sure if it's not too restrictive, but it reveals a great deal of places where we currently have uninitialized padding.

Could you wait until tomorrow, so I can push the current work?

**Comment by wesm** [comment]
I'm not sure I agree with that approach.

I have already done a lot of work this afternoon that I've just pushed. I think I'm close to having this fixed, can you give me a few hours?

File: cpp/src/arrow/builder.cc
Comment (wesm) [inline cpp/src/arrow/builder.cc]: @pcmoritz @alendit this is the memset that fixed the Python determinism unit test. Now I'm trying to figure out why

Comment (pcmoritz) [reply cpp/src/arrow/builder.cc]: Ok, so some of the initial bytes from the builder are leaking. One possibility I can think of is that we forget to call `Finish` somewhere in one of the builders in `python_to_arrow.cc` and just grab the array, I'm looking into that right now (but haven't found anything so far).

Comment (pcmoritz) [reply cpp/src/arrow/builder.cc]: @wesm Actually, the types_ builder in SequenceBuilder is never finished, could that be a problem? On the other hand, I think only bytes that have been written by us are extracted from there.

**Comment by wesm** [comment]
OK, the tests are passing. I don't understand why the memset is needed in `PrimitiveBuilder<T>::Init`. I can't spend any more time on this, so I hope someone can figure it out, but we should not block the 0.10.0 release on it, there are so many other things that need to get done before the release goes out.

Note I added memset in BooleanBuilder::Init/Resize -- these may not be necessary but removing them causes valgrind warnings in the tests. If we leave the memset, then we don't need the memsets in the AppendNull/AppendNulls functions.

File: cpp/src/arrow/builder.cc
Comment (wesm) [reply cpp/src/arrow/builder.cc]: Yep, that's the problem.

**Comment by wesm** [comment]
Seems like @pcmoritz found the problem in python_to_arrow.cc

**Comment by pcmoritz** [comment]
Ok, let me look into fixing this, I'll work on this PR.

**Comment by pcmoritz** [comment]
Ok, that seems to do the trick. I'm running the full Ray test suite against it again here: https://travis-ci.com/ray-project/ray/builds/78393000

If there is no more problems, the PR looks good to me.

**Comment by pcmoritz** [comment]
Ok, looks like it's working!

**Comment by alendit** [comment]
Oh, I should have mentioned the unfinished builders in serialization yesterday, sorry, timezones...

Having warnings on unzeroed data access turns up a couple more places, where we do a similar thing. We can merge this pr to fix plasma/ray and I'll address the other issues in further PRs.

**Comment by pcmoritz** [comment]
Yeah, that sounds good. Checking in the Builder destructor if Finish has ever been called might be a good idea. Do you want to open a JIRA for that?

**Comment by alendit** [comment]
Oh, I just pushed a minor change which came up in the review (another name for a variable and a simpler type size computation). I'll open another PR for them.

**Comment by wesm** [comment]
@pcmoritz I would have liked to have a chance to review this again before being merged; I will be better saying "please do not merge yet" in the future. I want to do some more work on this (to try to remove the other memset calls) so I will submit another patch

**Pull request closed without merging**
